src/Groups/matrices/matrix_manipulation.jl

```diff
@@ -0,0 +1,276 @@
+"""
+function complement(V::AbstractAlgebra.Generic.FreeModule{T}, W::AbstractAlgebra.Generic.Submodule{T}) where T <: FieldElem
+   @assert issubmodule(V,W) "The second argument is not a subspace of the first one"
+   if dim(W)==0 return sub(V,basis(V)) end
+
+   e = W.map
+
+   H = matrix( vcat([e(g) for g in gens(W)], [zero(V) for i in 1:(dim(V)-dim(W)) ]) )
+   d = dim(W)
+   A_left = identity_matrix(base_ring(V), dim(V))
+   A_right = identity_matrix(base_ring(V), dim(V))
+   for rn in 1:dim(W)     # rn = row number
+      cn = rn    # column number
+      while H[rn,cn]==0 cn+=1 end   # bring on the left the first non-zero entry
+      swap_cols!(H,rn,cn)
+      swap_rows!(A_right,rn,cn)
+      for j in rn+1:dim(W)
+         add_row!(H,H[j,rn]*H[rn,rn]^-1,rn,j)
+         add_column!(A_left,A_left[j,rn]*A_left[rn,rn]^-1,j,rn)
+      end
+   end
+   for j in dim(W)+1:dim(V)  H[j,j]=1  end
+   H = A_left*H*A_right
+   _gens = [V([H[i,j] for j in 1:dim(V)]) for i in dim(W)+1:dim(V) ]
+
+   return sub(V,_gens)
+end
+
+"""
+    permutation_matrix(F::Ring, Q::AbstractVector{T}) where T <: Int
+Return the permutation matrix over the ring `R` corresponding to `Q`. Here, `Q` must contain exactly once every integer from 1 to some `n`.
+"""
+function permutation_matrix(F::Ring, Q::AbstractVector{T}) where T <: Int
+   @assert Set(Q)==Set(1:length(Q)) "Invalid input"
+   Z = zero_matrix(F,length(Q),length(Q))
+   for i in 1:length(Q) Z[i,Q[i]] = 1 end
+   return Z
+end
+
+"""
+    evaluate(f::PolyElem, X::MatElem)
+Evaluate the polynomial `f` in the matrix `X`.
+"""
+evaluate(f::PolyElem, X::MatElem) = sum([X^i*base_ring(X)(coeff(f,i)) for i in 0:degree(f)])
+
+########################################################################
+#
+# New properties
+#
+########################################################################
+
+# TODO: not sure whether this definition of skew-symmetric is standard (for fields of characteristic 2)
+"""
+    isskewsymmetric_matrix(B::MatElem{T}) where T <: Ring
+Return whether the matrix `B` is skew-symmetric, i.e. `B = -transpose(B)` and `B` has zeros on the diagonal. Returns `false` if `B` is not a square matrix.
+"""
+function isskewsymmetric_matrix(B::MatElem{T}) where T <: RingElem
+   n = nrows(B)
+   n==ncols(B) || return false
+
+   for i in 1:n
+      B[i,i]==0 || return false
+      for j in i+1:n
+         B[i,j]==-B[j,i] || return false
+      end
+   end
+
+   return true
+end
+
+"""
+    ishermitian_matrix(B::MatElem{T}) where T <: Ring
+Return whether the matrix `B` is hermitian, i.e. `B = conjugate_transpose(B)`. Returns `false` if `B` is not a square matrix, or the field has not even degree.
+"""
+function ishermitian_matrix(B::MatElem{T}) where T <: RingElem
+   n = nrows(B)
+   n==ncols(B) || return false
+   e = degree(base_ring(B))
+   iseven(e) ? e = div(e,2) : return false
+
+   for i in 1:n
+      for j in i:n
+         B[i,j]==frobenius(B[j,i],e) || return false
```

Comment:
In this case, the "conjugation" it refers to is the involutory automorphism of the finite field of even degree, so the `conjugate_transpose` applies the automorphism to all entries of the matrix and then transpose the matrix (or viceversa, the two procedures commute each other). I often encountered this nomenclature during my PhD, but I don't know if it is not standard in other environments.

Reply:
The notation is absolutely standard for matrices over a a ring-with-involution. The problem was that you are hardwiring it to only work for matrices over finite-fields-with-involution. So if I wanted to have a method of this for other baserings, I couldn't do it, because you were blocking the general signature. Now you restricted it to `FinFieldElem`, so that should be fine now.

The same applies to other places using frobenius, though -- in this PR, that'd be `conjugate_transpose`, which should also be restricted to `FinFieldElem`

Reply:
I can turn the input type into `FinFieldElem` whenever the involved operations deal only with finite fields.